src/Perpetuum.RequestHandlers/Zone/Containers/ZoneChangeModule.cs

```diff
@@ -1,49 +1,76 @@
-using Perpetuum.Robots;
+using Perpetuum.Players;
 using Perpetuum.Zones;
 
 namespace Perpetuum.RequestHandlers.Zone.Containers
 {
-    public class ZoneChangeModule : ZoneContainerRequestHandler
+    public abstract class ZoneChangeModule : ZoneContainerRequestHandler
     {
+        /// <summary>
+        /// Equip or Remove module based on implementing class
+        /// Must be called inside of transaction in HandleRequest
+        /// </summary>
+        public abstract void DoChange(IZoneRequest request, Player player, Container container);
+
         public override void HandleRequest(IZoneRequest request)
         {
             using (var scope = Db.CreateTransaction())
             {
                 var character = request.Session.Character;
                 var player = request.Zone.GetPlayerOrThrow(character);
 
+                //Store ratios before transaction
+                var hpRatio = player.ArmorPercentage;
+                var apRatio = player.CorePercentage;
+
                 CheckPvpState(player).ThrowIfError();
                 CheckCombatState(player).ThrowIfError();
                 CheckActiveModules(player).ThrowIfError();
 
-                player.EnlistTransaction();
+                var containerEid = request.Data.GetOrDefault<long>(k.containerEID);
+                var container = request.Zone.FindContainerOrThrow(player, containerEid);
 
-                var componentType = request.Data.GetOrDefault<string>(k.sourceComponent).ToEnum<RobotComponentType>();
-                var component = player.GetRobotComponentOrThrow(componentType);
-                var sourceSlot = request.Data.GetOrDefault<int>(k.source);
-                var targetSlot = request.Data.GetOrDefault<int>(k.target);
-                component.ChangeModuleOrThrow(sourceSlot, targetSlot);
+                CheckContainerType(container).ThrowIfError();
+                CheckFieldTerminalRange(player, container).ThrowIfError();
 
-                player.Initialize(character);
 
-                player.Save();
+                player.EnlistTransaction();
+                container.EnlistTransaction();
 
-                Transaction.Current.OnCommited(() =>
+                Transaction.Current.OnCompleted(completed =>
                 {
+                    player.Initialize(character);
+
+                    //Apply original ratios after unequip-action
+                    player.Armor = player.ArmorMax * hpRatio;
+                    player.Core = player.CoreMax * apRatio;
+
                     player.SendRefreshUnitPacket();
 
                     var result = new Dictionary<string, object>
                     {
-                        { k.robot, player.ToDictionary() }
+                        {k.robot, player.ToDictionary()},
+                        {k.container, container.ToDictionary()}
                     };
                     Message.Builder.FromRequest(request).WithData(result).WrapToResult().Send();
                 });
```

Comment:
Be aware that there is no guarantee this code is run at this point in the method. This is a callback delegate and will only be executed once the OnCompleted event is fired. The code below this section will run immediately and it may not give you the results you want, especially under heavy load.
In order to be more explicit about this, I recommend making this a separate method that returns `void` so that anyone looking at this code is under no illusions that the Armor and Core modifications are discrete modifications outside of the scope.

Signature should be similar to `private static void OnCompleted(Player player, Message message)` and called via `Transaction.Current.OnCompleted(completed => OnCompleted(player, Message));`

This will prevent an assumption that the code has been executed prior to the `DoChange(request, player, container);` and other player based methods.